mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java

```diff
@@ -0,0 +1,127 @@
+    this.mainDeserializer = deserializer(schema.asStruct(), fieldInspector);
+  }
+
+  Record deserialize(Object data) {
+    return (Record) mainDeserializer.value(data);
+  }
+
+  private interface FiledDeserializer {
+    Object value(Object object);
+  }
+
+  private static FiledDeserializer deserializer(Type type, ObjectInspector fieldInspector) throws SerDeException {
+    switch (type.typeId()) {
+      case BOOLEAN:
+        return o -> ((BooleanObjectInspector) fieldInspector).get(o);
+      case INTEGER:
+        return o -> ((IntObjectInspector) fieldInspector).get(o);
+      case LONG:
+        return o -> ((LongObjectInspector) fieldInspector).get(o);
+      case FLOAT:
+        return o -> ((FloatObjectInspector) fieldInspector).get(o);
+      case DOUBLE:
+        return o -> ((DoubleObjectInspector) fieldInspector).get(o);
+      case STRING:
+        return o -> ((StringObjectInspector) fieldInspector).getPrimitiveJavaObject(o);
+      case UUID:
+        // TODO: This will not work with Parquet. Parquet UUID expect byte[], others are expecting UUID
+        return o -> UUID.fromString(((StringObjectInspector) fieldInspector).getPrimitiveJavaObject(o));
+      case DATE:
+      case TIMESTAMP:
+      case FIXED:
+      case BINARY:
+      case DECIMAL:
+        // Iceberg specific conversions
+        return o -> ((IcebergWriteObjectInspector) fieldInspector).getIcebergObject(o);
+      case STRUCT:
+        return new StructDeserializer((Types.StructType) type, (StructObjectInspector) fieldInspector);
+      case LIST:
+      case MAP:
+      case TIME:
+      default:
+        throw new SerDeException("Unsupported column type: " + type);
+    }
+  }
+
+  private static class StructDeserializer implements FiledDeserializer {
+    private final FiledDeserializer[] filedDeserializers;
+    private final StructObjectInspector fieldInspector;
+    private final Types.StructType type;
+
+    private StructDeserializer(Types.StructType type, StructObjectInspector fieldInspector) throws SerDeException {
+      List<? extends StructField> structFields = fieldInspector.getAllStructFieldRefs();
+      List<Types.NestedField> nestedFields = type.fields();
+      this.filedDeserializers = new FiledDeserializer[structFields.size()];
+      this.fieldInspector = fieldInspector;
+      this.type = type;
+
+      for (int i = 0; i < filedDeserializers.length; i++) {
+        filedDeserializers[i] =
+            deserializer(nestedFields.get(i).type(), structFields.get(i).getFieldObjectInspector());
+      }
+    }
+
+    @Override
+    public Record value(Object object) {
+      if (object == null) {
+        return null;
+      }
+
+      List<Object> data = fieldInspector.getStructFieldsDataAsList(object);
+      Record result = GenericRecord.create(type);
+
+      for (int i = 0; i < filedDeserializers.length; i++) {
+        Object fieldValue = data.get(i);
+        if (fieldValue != null) {
```

Comment:
if the fieldValue is null, shouldn't we still set the null in the result record?

Reply:
We created a new Record here, so it should be null by default.
That said it should not hurt too much on performance, and easier to read when this is explicit, so added the extra set